Feat: accept cron schedule for analyzers

murilommen opened the pull request:
Previously we had left out the ability to define an `Analyzer.schedule = CronSchedule`. This PR adds that as a feature and also includes two guardrails for the cron string:
- it must respect 5 slots (and not 6+)
- it must be greater than or equal to 1hour
I've used the `re.match` to capture invalid strings and also a unit test with a few different examples

murilommen committed:
accept cron schedule for analyzers

murilommen committed:
fix typing

christinedraper commented on whylabs_toolkit/monitor/models/analyzer/analyzer.py:
```diff
@@ -57,8 +60,8 @@ class Analyzer(NoExtrasBaseModel):
-    # disabling CronSchedule as it can be tricky on the BE
-    schedule: Optional[FixedCadenceSchedule] = Field(  # Optional[Union[CronSchedule, FixedCadenceSchedule]] = Field(
+
+    schedule: Optional[Union[FixedCadenceSchedule, CronSchedule]] = Field(
```
Are there any sideeffects to setting the schedule to None like this? In particular it would be better if the field wasnt included in a generated yaml as we dont particularly want people to start experimenting with it

murilommen replied:
by default we'll fill in with a `FixedCadenceSchedule` if the user doesn't set anything in the toolkit. if they explicitly unset the schedule, the analyzer is considered disabled by the monitor. do you think we should not allow it at all?

christinedraper replied:
its fine if we normally default it to fixed. I forgot it's always expected

christinedraper commented on whylabs_toolkit/monitor/models/analyzer/analyzer.py:
```diff
@@ -23,6 +24,8 @@
 from .targets import ColumnMatrix, DatasetMatrix
 
+CRON_REGEX = "^(0|\d) (\d+|\d+-\d+|\/\d+)(,\d+|\d+-\d+|\/\d+)* (\*|\d+|\d+-\d+|\/\d+) (\*|\d+|\d+-\d+|\/\d+) (\*|\d+|\d+-\d+|(1|2|3|4|5|6|0)(,\d+)*)$"
+
```
seems like this only supports comma format for days - is it not valid for the other fields too? Maybe should use something like croniter to check its valid?

murilommen replied:
I prefer not to include a new dependency, so i've changed everything to be checked for with python. see if it's enough for now.

murilommen committed:
change from regex to python for validation

christinedraper commented on whylabs_toolkit/helpers/cron_validators.py:
```diff
@@ -0,0 +1,53 @@
+def _is_not_less_granular_than_1_hour(split_cron: SplitCron) -> bool:
+    """Check if the cron expression is less granular than 1 hour."""
+    # Specific days checks
+    if split_cron.minute != "*" and split_cron.minute != "0":
```
I dont get why it needs to check that its not 0 (and ditto for the other fields). Seems like this check should fail if the minute is '*' or anything with a - or  , in it. Anything else would be ok.

murilommen replied:
true. it really doesn't make sense to check for the "top of the {time}"

christinedraper commented on whylabs_toolkit/helpers/cron_validators.py:
```diff
@@ -0,0 +1,53 @@
+    if split_cron.day_of_week != "*" and split_cron.day_of_week != "1":
+        return True
+
+    # Check range
```
all that happens from here on is that it iterates and then returns False... logic seems wrong. Also seems disconnected from "is granularity less than hour", seems like it is meant to be checking any range fields

murilommen committed:
cron expression validation fixes

details:
- include every N {field} on base regex on CronSchedule
- get rid of "strictly 0 {field}" cases, which didn't make sense
- fix order of SplitCron implementation
- checks for "," and "-" cases only on the minute field
- added tests

murilommen commented:
@christinedraper added a few changes that addresses these comments, see if they're enough to resolve the conversations 🙏🏽 and thanks for catching these

christinedraper reviewed (approved):
Is CRON_REGEX still used? I didnt try to validate it

murilommen commented:
> Is CRON_REGEX still used? I didnt try to validate it

It is used when creating CronSchedule and pydantic applies it
